### PR TITLE
add shuffle action

### PR DESCRIPTION
**Pull request by Thor-wl**
ref: https://github.com/volcano-sh/volcano/issues/1777
design doc: https://github.com/volcano-sh/volcano/blob/master/docs/design/rescheduling.md
Signed-off-by: Thor-wl <13164644535@163.com>

### CODE REVIEW

**Comment by k82cn** [comment]
> This PR should work together with https://github.com/volcano-sh/volcano/pull/2094 https://github.com/volcano-sh/volcano/pull/2095 https://github.com/volcano-sh/volcano/pull/2096. So please DO NOT merge it until the above PR are merged.

Why does this PR have such kind of dependencies?

**Comment by k82cn** [comment]
Please help to build UT for this action instead of reusing rescheduling UT.

File: pkg/scheduler/api/types.go [hunk 1]
```diff
@@ -171,3 +171,6 @@ type VictimTasksFn func() []*TaskInfo
 
 // AllocatableFn is the func declaration used to check whether the task can be allocated
 type AllocatableFn func(*QueueInfo, *TaskInfo) bool
+
+// VictimsFromCandidatesFn is the func declaration used to select victim tasks from candidates and evict them
+type VictimsFromCandidatesFn func([]*TaskInfo) []*TaskInfo
```
Comment (k82cn) [inline pkg/scheduler/api/types.go:176]: Why not reuse `VictimTasksFn`?

File: pkg/scheduler/actions/shuffle/shuffle_test.go [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+	schedulingv1 "k8s.io/api/scheduling/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+
+	schedulingv1beta1 "volcano.sh/apis/pkg/apis/scheduling/v1beta1"
+	"volcano.sh/volcano/pkg/scheduler/api"
+	"volcano.sh/volcano/pkg/scheduler/cache"
+	"volcano.sh/volcano/pkg/scheduler/conf"
+	"volcano.sh/volcano/pkg/scheduler/framework"
+	"volcano.sh/volcano/pkg/scheduler/plugins/rescheduling"
```
Comment (k82cn) [inline pkg/scheduler/actions/shuffle/shuffle_test.go:33]: shuffle action should not dependent on rescheduling plugin.

File: pkg/scheduler/api/types.go (hunk 1 above)
Comment (Thor-wl) [reply pkg/scheduler/api/types.go:176]: `VictimTaskFn` is a function without input parameters, which is not so suitable for shuffle action. In shuffle action, the input parameter should be a collection of running tasks.

File: pkg/scheduler/actions/shuffle/shuffle_test.go (hunk 2 above)
Comment (Thor-wl) [reply pkg/scheduler/actions/shuffle/shuffle_test.go:33]: Just as `allocate_test.go` and `preempt_test.go` works, all the UT of actions should work with some registered plugins and then get an expected output. So this UT registers `rescheduling` plugin.

File: pkg/scheduler/api/types.go (hunk 1 above)
Comment (k82cn) [reply pkg/scheduler/api/types.go:176]: That should be ok to update `victimTasksFn` for that.

File: pkg/scheduler/actions/shuffle/shuffle_test.go (hunk 2 above)
Comment (k82cn) [reply pkg/scheduler/actions/shuffle/shuffle_test.go:33]: Let's build a mock plugin for action UT, and we should also update allocate_test & preempt_test to have mock plugins.

Comment (Thor-wl) [reply pkg/scheduler/actions/shuffle/shuffle_test.go:33]: It is reasonable in some degree. Perhaps we can submit another PR especially for that.

Comment (k82cn) [reply pkg/scheduler/actions/shuffle/shuffle_test.go:33]: > It is reasonable in some degree. Perhaps we can submit another PR especially for that.

How to handle CI error?

File: pkg/scheduler/api/types.go (hunk 1 above)
Comment (shinytang6) [reply pkg/scheduler/api/types.go:176]: +1 for resuing VictimTasksFn

Comment (Thor-wl) [reply pkg/scheduler/api/types.go:176]: OK. Let me give a upate.

**Comment by william-wang** [comment]
> ref: #1777 design doc: https://github.com/volcano-sh/volcano/blob/master/docs/design/rescheduling.md Signed-off-by: Thor-wl [13164644535@163.com](mailto:13164644535@163.com)

ref: #1955
Associate this pr to issue #1955. the shuffle action will be a seperated common action for multiple plugins.

**Pull request closed without merging**